src/internal/connector/onedrive/restore.go

```diff
@@ -118,50 +146,130 @@ func RestoreCollection(
 		select {
 		case <-ctx.Done():
 			errUpdater("context canceled", ctx.Err())
-			return metrics, true
+			return metrics, folderPerms, true
 
 		case itemData, ok := <-items:
 			if !ok {
-				return metrics, false
-			}
-			metrics.Objects++
-
-			metrics.TotalBytes += int64(len(copyBuffer))
-
-			itemInfo, err := restoreItem(ctx,
-				service,
-				itemData,
-				drivePath.DriveID,
-				restoreFolderID,
-				copyBuffer,
-				source)
-			if err != nil {
-				errUpdater(itemData.UUID(), err)
-				continue
+				return metrics, folderPerms, false
 			}
 
 			itemPath, err := dc.FullPath().Append(itemData.UUID(), true)
 			if err != nil {
 				logger.Ctx(ctx).DPanicw("transforming item to full path", "error", err)
 				errUpdater(itemData.UUID(), err)
-
 				continue
 			}
 
-			deets.Add(
-				itemPath.String(),
-				itemPath.ShortRef(),
-				"",
-				true,
-				itemInfo)
+			if source == OneDriveSource {
+				name := itemData.UUID()
+				if strings.HasSuffix(name, ".data") {
+					metrics.Objects++
+					metrics.TotalBytes += int64(len(copyBuffer))
+					trimmedName := name[:len(name)-5]
+
+					itemID, itemInfo, err = restoreData(ctx, service, trimmedName, itemData, drivePath.DriveID, restoreFolderID, copyBuffer, source)
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					restoredIDs[trimmedName] = itemID
+
+					deets.Add(itemPath.String(), itemPath.ShortRef(), "", true, itemInfo)
+				} else if strings.HasSuffix(name, ".meta") {
+					meta, err := getMetadata(itemData.ToReader())
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					trimmedName := name[:len(name)-5]
+					restoreID, ok := restoredIDs[trimmedName]
+					if !ok {
+						errUpdater(itemData.UUID(), fmt.Errorf("item not available to restore permissions"))
+						continue
+					}
+
+					restorePermissions(ctx, service, drivePath.DriveID, restoreID, parentPerms, meta.Permissions)
+
+					// TODO(meain): Finalize on how we count success
+					metrics.Successes++ // Counted as success when we have also restored metadata
```

Comment:
As of now, we add `metrics.Object` when we encounter a new `.data` file and we consider it as success when we have also been able to restore its permissions. LMK if this needs any changes.